### ISSUE TITLE
Full asm!() support

### ISSUE THREAD

**Issue by bjorn3** [bjorn3/rustc_codegen_cranelift#1204]
* [ ] Support `const` operands.
* [ ] Support `sym` operands.
* [ ] Support register classes. This will require writing a mini register allocator to solve all register constraints.
* [ ] Optimize
    * [ ] Overlap input and output stack slots.
    * [ ] Skip saving caller saved registers.
* [ ] Support more architectures
    * [ ] 32bit x86
    * [ ] arm32
    * [ ] aarch64
    * [ ] riscv32/64
    * [ ] ...

Tracking issue for stabilizing `asm!()`: https://github.com/rust-lang/rust/issues/72016

**Comment by bjorn3** [comment]
> Avoid the object file linking hack. It doesn't work on macOS and Windows but is currently necessary as incr comp expects each codegen unit to provide a single object file instead of allowing an extra object file for the compiled assembly.

I managed to do this on the parallel_comp_refactor branch.

**Comment by bjorn3** [comment]
Merged in #1264.

**Comment by bjorn3** [comment]
Mostly implemented const and sym operands for both inline and global asm.

**Comment by bjorn3** [comment]
As of https://github.com/rust-lang/rustc_codegen_cranelift/pull/1403 inline asm is now supported and stable on all targets supported by cg_clif.

**Comment by repi** [comment]
awesome!

did a quick test with building our codebase with this and did fail on `wasmtime-runtime` using inline asm:

```
error: asm! and global_asm! sym operands are not yet supported
   --> /home/repi/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmtime-runtime-14.0.2/src/libcalls.rs:156:5
    |
156 |     wasmtime_environ::foreach_builtin_function!(libcall);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this error originates in the macro `std::arch::global_asm` which comes from the expansion of the macro `wasmtime_environ::foreach_builtin_function` (in Nightly builds, run with -Z macro-backtrace for more info)
```

related to this note?

> For global asm (https://github.com/rust-lang/rustc_codegen_cranelift/pull/1351, but doesn't work when the function is made private to the codegen unit)

or maybe this was not part of the new deployed rustup component?

**Comment by bjorn3** [comment]
> related to this note?

Correct. So far I had only checked wasmtime with unstable features enabled (as is the default for building it using `./y.sh` in this repo), but not with unstable features disabled (as is the case for the rustc-codegen-cranelift-preview rustup component) With unstable features enabled the entire wasmtime test suite passes with the exception of a couple of tests that need panic=unwind.

**Comment by repi** [comment]
ah good to know, and fantastic that the full test suite passes!
